### ISSUE TITLE
DiscreteBelief and DiscreteUpdater for Finite Horizon POMDPs

### ISSUE THREAD

**Issue by Omastto1** [JuliaPOMDP/BeliefUpdaters.jl#22]
The main problem with `DiscreteBelief` is that its methods do not support the `stage` argument which the Finite Horizon POMDPs need. At this point, I think we need to somehow differ between `DiscreteBelief` for Infinite and  Finite POMDPs. Defining a new structure for Finite Horizon POMDPs would be from 99% just copying the code from the Infinite Horizon ones. Maybe some sort of dynamic dispatch could solve this?

The methods that I updated for Finite Horizons are `pdf` `initialize_belief` and `update` as they all work with for example `states` but FInite Horizon POMDPs need `stage_states`.

In `pdf` I

**Comment by zsunberg** [comment]
I agree that `DiscreteUpdater` could be made faster for the finite horizon case. You can implement this with the `HorizonLength` trait. i.e.

```
initialize_belief(up::DiscreteUpdater, d) = initialize_beilief(HorizonLength(up.pomdp), up, d)
initialize_belief(::FiniteHorizon, up::DiscreteUpdater, d) = StageDiscreteBelief(#=construct this with whatever stage the states in d have=#)
initialize_belief(::InfiniteHorizon, up::DiscreteUpdater, d) = DiscreteBelief(#=construct this exactly like it is in the code now=#)
```

You can find more about traits here: https://docs.julialang.org/en/v1/manual/methods/#Trait-based-dispatch

**Comment by Omastto1** [comment]
Ok, I got a significant part implemented, now I am adding some testing.

However to be clear, at this moment I used trait-based-dispatch, but even in Your code, You are using `StagedDiscreteBelief`.

If we want to create a new class specifically for staged beliefs I believe that the trait-based dispatch is redundant over dynamic dispatch.
Also, I came over a couple of problems. First, with trait-based dispatch, it is not possible to define `uniform_belief` as the method does not accept the `stage` parameter. Second `DiscreteBelief` can not store `stage`. Right now, I am not sure whether those are really problems as I did not need to use any of discussed methods in a staged scenarious, but I think `uniform_belief` _could_ be useful.

Also if we want to add `StagedDiscreteBelief`, do we add `AbstractDiscreteBelief` as well?

**Comment by zsunberg** [comment]
Sorry for the delayed response! I was on a camping trip this weekend.

> If we want to create a new class specifically for staged beliefs I believe that the trait-based dispatch is redundant over dynamic dispatch.

How do you implement `initialize_belief` without trait-based dispatch?

> If we want to create a new class specifically for staged beliefs I believe that the trait-based dispatch is redundant over dynamic dispatch.

I assume by "new class" you mean `StagedDiscreteBelief`. I agree that you could skip the trait-based dispatch for `update`, but I think you still want it for `initialize_belief`.

> it is not possible to define uniform_belief as the method does not accept the stage parameter.

Since `uniform_belief` is just a convenience function and not part of the core POMDPs.jl interface, you probably don't need to implement it. If you do decide you need to implement it, would it make sense to define it as a uniform belief over all stage 1 states?

> Also if we want to add StagedDiscreteBelief, do we add AbstractDiscreteBelief as well?

Hmmm... If you find yourself using it for dispatch, we can add it, but I honestly haven't found abstract types all that useful. It's up to you.

**Comment by Omastto1** [comment]
I hope you enjoyed your trip!

I am still unsure whether to use one class, `DiscreteBelief` and trait-based dispatch, or define `StagedDiscreteBelief`, which would also store the `stage` information. 

However, the only Finite Horizon POMDP algorithm implemented (still not part of POMDPs.jl, unfortunately) stores beliefs belonging to a specific stage in a separate array, thus removing the need to store `stage` info in belief explicitly.  Making the `StagedDiscreteBelief` redundant, for now.

**Comment by Omastto1** [comment]
Well, I have been playing with tests for this and found out the following.

To check the correctness of the Finite Horizon POMDPs algorithm, I solve the Finite Horizon problem in Infinite Horizon solver.
In this case, the Infinite horizon solver (PBVI) awaits `DiscreteBelief` of **all** possible states over all stages. However, I am solving the Finite Horizon Problem, which results in `DiscreteBelief`s containing only states of a given stage.

This Results in PBVI failing as it awaits a whole state space. 

We eliminate this problem (where someone wants to solve Finite Horizon Problems defined with `fixhorizon()`), if we, instead of trait-based dispatch built on top of HorizonLength() function, define dynamic dispatch methods and `DiscreteBelief` and `StagedBelief`. Infinite Horizon Solvers will use `DiscreteBeliefs` and the Finite horizon solvers `StagedDiscreteBelief`s.

The question is whether this problem is good enough to define a new class. There is no real-world purpose for solving `fixhorizon()` problems with Infinite Horizon solvers, but I think it is a good illustration of Finite Horizon solvers' performance.

**Comment by zsunberg** [comment]
> To check the correctness of the Finite Horizon POMDPs algorithm, I solve the Finite Horizon problem in Infinite Horizon solver. In this case, the Infinite horizon solver (PBVI) awaits DiscreteBelief of all possible states over all stages. However, I am solving the Finite Horizon Problem, which results in DiscreteBeliefs containing only states of a given stage.

This is **very** important for interoperability. From the perspective of the distribution interface, **all** beliefs should act like distributions over **all** states. For finite horizon problems, we may want to construct a special sparse belief (i.e. `StagedDiscreteBelief`) that takes advantage of the fact that we only need to **store** numbers for one stage, but it should still work with the policy from an infinite horizon solver. This is like how a sparse matrix acts like a dense matrix when you multiply it even though the storage of the numbers is different.

> We eliminate this problem (where someone wants to solve Finite Horizon Problems defined with fixhorizon()), if we, instead of trait-based dispatch built on top of HorizonLength() function, define dynamic dispatch methods and DiscreteBelief and StagedBelief. Infinite Horizon Solvers will use DiscreteBeliefs and the Finite horizon solvers StagedDiscreteBeliefs.

Both policies from infinite horizon solvers **and** finite horizon solvers should be able to work with `DiscreteBelief`s **and** `StagedDiscreteBelief`s.

> There is no real-world purpose for solving fixhorizon() problems with Infinite Horizon solvers,

I think there **is** a real-world reason: the implementation of online tree-search solvers is exactly the same for fixed and infinite horizon problems, and we only want to write an infinite horizon version of those solvers, but it should still work with a finite horizon problem created with `fixhorizon()`.

> However, the only Finite Horizon POMDP algorithm implemented (still not part of POMDPs.jl, unfortunately) stores beliefs belonging to a specific stage in a separate array, thus removing the need to store stage info in belief explicitly.

This is a minor style note: Within the `solve` function, it seems like it would just be easier to use belief vectors rather than `DiscreteBelief` objects.

It might be easier to communicate if you submit a draft pull request so that I can comment on it.

**Comment by zsunberg** [comment]
Another concept to keep in mind is this:

A finite horizon POMDP *is* an infinite horizon POMDP where the state space includes the time step.

I think that might help you to understand how the beliefs and solvers should work together.

**Comment by Omastto1** [comment]
Sorry for the delayed response from my side this time. I helped to harvest the crops and haven't had time to solve this.
Anyway, I created a pull request so we can discuss more specific issues.
